### PR TITLE
[build] warning as errors

### PR DESCRIPTION
**Pull request by vvlevchenko**

### CODE REVIEW

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt [hunk 1]
```diff
@@ -15,6 +15,7 @@ import java.lang.IllegalStateException
  *
  * [omitEmptyLines] is useful for testing output (e.g. diff calculating).
  */
+@Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
```
Comment (olonho) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt]: Why not fix properly by rename parameter?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt [hunk 2]
```diff
@@ -163,6 +163,7 @@ private fun getFeaturedLibraries(
 ) : List<KonanLibrary> {
     val remainingFeaturedLibraries = featuredLibraryFiles.toMutableSet()
     val result = mutableListOf<KonanLibrary>()
+    @Suppress("UNCHECKED_CAST")
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt]: Hmm, can't we fix the signature?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanDriver.kt [hunk 3]
```diff
@@ -25,6 +25,7 @@ fun runTopLevelPhases(konanConfig: KonanConfig, environment: KotlinCoreEnvironme
     if (konanConfig.infoArgsOnly) return
 
     try {
+        @Suppress("UNCHECKED_CAST")
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanDriver.kt]: Why this cast is needed at all?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt [hunk 4]
```diff
@@ -80,6 +81,7 @@ fun IrClass.isObjCProtocolClass(): Boolean =
 fun ClassDescriptor.isObjCProtocolClass(): Boolean =
         this.fqNameSafe == objCProtocolFqName
 
+@Suppress("DEPRECATION")
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt]: Can it be properly fixed?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/cgen/CBridgeGenUtils.kt [hunk 5]
```diff
@@ -20,7 +20,7 @@ import org.jetbrains.kotlin.ir.expressions.impl.IrTryImpl
 import org.jetbrains.kotlin.ir.symbols.impl.IrSimpleFunctionSymbolImpl
 import org.jetbrains.kotlin.ir.symbols.impl.IrValueParameterSymbolImpl
 import org.jetbrains.kotlin.ir.types.IrType
-import org.jetbrains.kotlin.ir.types.impl.IrUninitializedType
+import org.jetbrains.kotlin.ir.types.impl.*
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/cgen/CBridgeGenUtils.kt]: Why this change?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/NewIrUtils.kt [hunk 6]
```diff
@@ -67,6 +67,7 @@ val IrClass.isFinalClass: Boolean
 
 fun IrClass.isSpecialClassWithNoSupertypes() = this.isAny() || this.isNothing()
 
+@Suppress("UNCHECKED_CAST")
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/NewIrUtils.kt]: Cannot we make this cast checking, by swithcing to reified inline?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/DebugUtils.kt [hunk 7]
```diff
@@ -224,7 +225,7 @@ internal fun IrType.llvmType(context:Context): LLVMTypeRef = context.debugInfo.l
     }
 }
 
-internal fun IrType.encoding(context: Context): DwarfTypeKind = when(computePrimitiveBinaryTypeOrNull()) {
+internal fun IrType.encoding(): DwarfTypeKind = when(computePrimitiveBinaryTypeOrNull()) {
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/DebugUtils.kt]: val?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/cgen/CBridgeGenUtils.kt (hunk 5 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/cgen/CBridgeGenUtils.kt]: Because we can't hang suppression on import

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt (hunk 1 above)
Comment (vvlevchenko) [reply Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt]: the name of parameter doesn't corresponds the semantics of variable so that why I've suppressed this warning

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt (hunk 4 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt]: This method uses locally declared deprecated function. It'd be better to drop suppression because it's already covered by file:Suppress

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt [hunk 8]
```diff
@@ -1,3 +1,4 @@
+@file:Suppress("DEPRECATION")
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt]: Hmm, better suppress warnings case-by-case, not on file or class scope.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/CallableReferenceLowering.kt [hunk 9]
```diff
@@ -280,6 +280,7 @@ internal class CallableReferenceLowering(val context: Context): FileLoweringPass
                         else -> irBuiltIns.anyClass.owner.constructors.single()
                     }
                     +irDelegatingConstructorCall(superConstructor).apply {
+                        @Suppress("LABEL_NAME_CLASH")
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/CallableReferenceLowering.kt]: Can we add explicit label?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt (hunk 4 above)
Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt]: Usually file-level suppresses are not very good idea.

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt]: Btw, as far as I know there are a lot of descriptor-based functions **deprecated by mistake**: they are used in the code that is to be migrated to frontend checkers.

Comment (olonho) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt]: Shall we remove deprecations then?

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ObjCInterop.kt]: Yes. Generally we should remove deprecations from methods that are directly or indirectly used to decide about `reportCompilationError` usage.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanDriver.kt (hunk 3 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/KonanDriver.kt]: to make invokeToplevel accept Context

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt (hunk 2 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/FeaturedLibraries.kt]: no, but we can do the following:
 val libraries = resolvedLibraries.getFullList(null).filterIsInstance<KonanLibrary>()

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt (hunk 1 above)
Comment (sbogolepov) [reply Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt]: Scope of this suppress can be reduced to `printer` object.

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrBridgeBuilder.kt [hunk 10]
```diff
@@ -19,6 +19,7 @@ class BridgeBuilderResult(
 /**
  * Generates [NativeBridges] and corresponding function bodies and property accessors.
  */
+@Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
 class StubIrBridgeBuilder(
```
Comment (sbogolepov) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrBridgeBuilder.kt]: `bridgeGeneratingVisitor` can be marked instead.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/CallableReferenceLowering.kt (hunk 9 above)
Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/CallableReferenceLowering.kt]: no

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/ContextUtils.kt [hunk 11]
```diff
@@ -394,7 +395,7 @@ internal class Llvm(val context: Context, val llvmModule: LLVMModuleRef) {
                 .filter {
                     require(it is KonanLibrary)
                     (!it.isDefault && !context.config.purgeUserLibs) || imports.nativeDependenciesAreUsed(it)
-                } as List<KonanLibrary>
+                }.cast()
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/ContextUtils.kt:398]: Casts better be explicit (with explicit type arguments in `.cast<...>()`)

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 12]
```diff
@@ -716,6 +714,7 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
                 using(parameterScope) {
                     using(VariableScope()) {
                         recordCoverage(body)
+                        @Suppress("LABEL_NAME_CLASH")
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:717]: Maybe use different explicit labels instead? https://kotlinlang.org/docs/reference/returns.html#return-at-labels
```kotlin
using(parameterScope) usingParameterScope@{
    using(VariableScope()) usingVariableScope@{
```

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:717]: Same here: https://github.com/JetBrains/kotlin-native/pull/3858/files#diff-e1fcf524f424fa310ecf024a7a7b9958R283

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/BuiltInFictitiousFunctionIrClassFactory.kt [hunk 13]
```diff
@@ -301,7 +304,22 @@ internal class BuiltInFictitiousFunctionIrClassFactory(
...
+                @Suppress("DEPRECATION")
+                /* TODO: [PropertyDescriptor::isDelegated] is deprecated. */
+                IrPropertyImpl(offset,
+                        offset,
+                        memberOrigin,
+                        s,
+                        descriptor.name,
+                        descriptor.visibility,
+                        descriptor.modality,
+                        descriptor.isVar,
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/BuiltInFictitiousFunctionIrClassFactory.kt:316]: Two many boolean flag arguments, using argument names is highly desirable here.

Comment (vvlevchenko) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/BuiltInFictitiousFunctionIrClassFactory.kt:316]: Oh, this is the only not deprecated constructor :/

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/DescriptorToIrTranslationUtils.kt [hunk 14]
```diff
@@ -79,16 +79,18 @@ internal interface DescriptorToIrTranslationMixin {
...
 
     fun createConstructor(constructorDescriptor: ClassConstructorDescriptor): IrConstructor {
         val irConstructor = symbolTable.declareConstructor(constructorDescriptor) {
+            // TODO: [IrUninitializedType] is deprecated.
+            @Suppress("DEPRECATION")
             IrConstructorImpl(
                 SYNTHETIC_OFFSET, SYNTHETIC_OFFSET,
                 IrDeclarationOrigin.IR_EXTERNAL_DECLARATION_STUB,
-                it, IrUninitializedType, constructorDescriptor
+                    it, IrUninitializedType, constructorDescriptor
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/DescriptorToIrTranslationUtils.kt:93]: Weird indent?

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/Context.kt [hunk 15]
```diff
@@ -90,8 +92,13 @@ internal class SpecialDeclarationsFactory(val context: Context) {
...
                     DECLARATION_ORIGIN_FIELD_FOR_OUTER_THIS,
-                    descriptor,
-                    outerClass.defaultType
+                    IrFieldSymbolImpl(descriptor),
+                    descriptor.name,
+                    outerClass.defaultType,
+                    descriptor.visibility,
+                    !descriptor.isVar,
+                    descriptor.isEffectivelyExternal(),
+                    descriptor.dispatchReceiverParameter == null
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/Context.kt:101]: Named arguments here too.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt [hunk 16]
```diff
@@ -713,8 +711,8 @@ internal class CodeGeneratorVisitor(val context: Context, val lifetimes: Map<IrE
                 declaration.location(start = false)) {
             using(FunctionScope(declaration, it)) {
                 val parameterScope = ParameterScope(declaration, functionGenerationContext)
-                using(parameterScope) {
-                    using(VariableScope()) {
+                using(parameterScope) usingParameterScope@ {
+                    using(VariableScope()) usingVariableScope@ {
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:715]: Is this formatted according to [Kotlin coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)?

Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/IrToBitcode.kt:715]: Also please recheck `return@using` below.

**Comment by vvlevchenko** [comment]
> I'm ok with changes, but they should be made also in main repo of library https://github.com/Kotlin/kotlinx-cli

https://github.com/Kotlin/kotlinx-cli/pull/38

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/DescriptorToIrTranslationUtils.kt [hunk 17]
```diff
@@ -79,12 +79,14 @@ internal interface DescriptorToIrTranslationMixin {
                 is PropertyDescriptor -> createProperty(it)
                 is FunctionDescriptor -> createFunction(it, IrDeclarationOrigin.FAKE_OVERRIDE)
                 else -> error("Unexpected fake override descriptor: $it")
-            } as IrDeclaration // Assistance for type inference.
+            } // Assistance for type inference.
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/DescriptorToIrTranslationUtils.kt:82]: The comment was describing the cast, so it should be removed.
